controllers/etcd_controller.go

```diff
@@ -1123,11 +1197,111 @@ func (r *EtcdReconciler) waitUntilStatefulSetReady(sts *appsv1.StatefulSet) (*ap
 			return gardenerretry.MinorError(err)
 		}
+		if ss.Status.Replicas != *ss.Spec.Replicas {
+			return gardenerretry.MinorError(fmt.Errorf("statefulset not scaled"))
+		}
 		return gardenerretry.Ok()
 	})
 	return ss, err
 }
 
+func (r *EtcdReconciler) getStatefulSet(ctx context.Context, etcd *druidv1alpha1.Etcd) (*appsv1.StatefulSet, error) {
+	selector, err := metav1.LabelSelectorAsSelector(etcd.Spec.Selector)
+	if err != nil {
+		logger.Error(err, "error converting etcd selector to labelSelector")
+		return nil, err
+	}
+
+	statefulSets := &appsv1.StatefulSetList{}
+	if err = r.List(ctx, statefulSets, client.InNamespace(etcd.Namespace), client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		return nil, err
+	}
+
+	if len(statefulSets.Items) > 0 {
+		return &statefulSets.Items[0], nil
+	}
+	return nil, nil
+}
+
+func (r *EtcdReconciler) getService(ctx context.Context, etcd *druidv1alpha1.Etcd) (*v1.Service, error) {
+	selector, err := metav1.LabelSelectorAsSelector(etcd.Spec.Selector)
+	if err != nil {
+		logger.Error(err, "error converting etcd selector to labelSelector")
+		return nil, err
+	}
+
+	services := &corev1.ServiceList{}
+	if err := r.List(ctx, services, client.InNamespace(etcd.Namespace), client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		logger.Error(err, "error listing services")
+		return nil, err
+	}
+
+	if len(services.Items) > 0 {
+		return &services.Items[0], nil
+	}
+	return nil, nil
+}
+
+func (r *EtcdReconciler) listPersistentVolumeClaims(ss *appsv1.StatefulSet) (*corev1.PersistentVolumeClaimList, error) {
+	selector, err := metav1.LabelSelectorAsSelector(ss.Spec.Selector)
+	if err != nil {
+		logger.Error(err, "error converting statefulset selector to selector")
+		return nil, err
+	}
+
+	persistentVolumeClaimList := &corev1.PersistentVolumeClaimList{}
+	if err := r.List(context.TODO(), persistentVolumeClaimList, client.InNamespace(ss.Namespace), client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		return nil, err
+	}
+
+	return persistentVolumeClaimList, nil
+}
+
+func (r *EtcdReconciler) prepareEtcdForScaleDown(etcd *druidv1alpha1.Etcd, ss *appsv1.StatefulSet, svc *v1.Service) (*corev1.PersistentVolumeClaimList, error) {
+	// TODO: wait for etcd health to be false so that etcd is not serving traffic
+
+	ctx, cancel := context.WithTimeout(context.TODO(), DefaultTimeout)
+	defer cancel()
+	if err := r.takeFullSnapshot(ctx, etcd, svc); err != nil {
+		return nil, err
+	}
+
+	pvcList, err := r.listPersistentVolumeClaims(ss)
+	if err != nil {
+		return nil, err
+	}
+
+	return pvcList, nil
+}
+
+func (r *EtcdReconciler) removePVCs(pvcList []corev1.PersistentVolumeClaim) error {
+	for _, pvc := range pvcList {
+		if err := r.Delete(context.TODO(), &pvc); err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+		ctx, cancel := context.WithTimeout(context.TODO(), DefaultTimeout)
+		defer cancel()
+		if err := r.waitForPVCDeletion(ctx, pvc.Name, pvc.Namespace); err != nil {
```

Comment:
Isn't it better to fire all deletes and then `RequeueAfter` for the errors?

Reply:
Yes, makes sense. Let me make that change as well. Thanks.